pkg/resource/v1/certconfig/resource.go

```diff
@@ -0,0 +1,58 @@
+package certconfig
+
+import (
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/framework"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "certconfigv1"
+)
+
+// Config represents the configuration used to create a new cloud config resource.
+type Config struct {
+	K8sClient kubernetes.Interface
+	Key       Key
+	Logger    micrologger.Logger
+}
+
+// Resource implements the cloud config resource.
+type Resource struct {
+	k8sClient kubernetes.Interface
+	key       Key
```

Comment:
When implementing this resource, instead of `key.ClusterID(obj)` there will now be `r.key.ClusterID(obj)` - I feel this is not a big change to readability.

Reply:
I am wondering why we need to do this anyway. What is wrong with having a key package for this here and just use it? All this is versioned and when something changes we have to change the version and its implementation anyway. So why having an interface now and injecting the key functions? I do not see the benefit somehow. This is also why I think we should just put the certconfig magic into the usual place and if necessary copy it where needed.

Reply:
I think functions will certconfig/key will be ortogonal to those in e.g. awsclusterconfig/key so this interface may be not that useful after all.

Do you have an example what could be different or repeated?

Reply:
Separation of concerns. This resource implementation is not tied to specific CRD type and therefore it cannot directly get information from given custom object. When operatorkit calls e.g. `GetDesiredState`, the implementation doesn't know what is type of customObject or how one can get e.g. `ClusterID` from it (the original reason to introduce `key` package). By introducing interface and requiring implementation of it as dependency this implementation is now oblivious to CRD type enabling this to be used from any CRD framework that just satisfies `Key` interface.

Reply:
> This resource implementation is not tied to specific CRD type 

This is my point. It should. This is why we do all the version magic and be so strict with all of this.

Reply:
I'm personally seeing versioning the other way round. E.g. `KVMClusterConfig` v1 imports interface and implementation for v1 certconfig. It provides implementation for custom object information extraction. If anything changes in `KVMClusterConfig` it can change its implementation without affecting certconfig version. Also works other way round, if certconfig changes anything internal, a new version is created and implementation that uses it upgrades when suitable. Please, correct me if I'm wrong, but internal implementation in certconfig shouldn't depend on how e.g. `KVMConfig` is implemented. Also works other way round.

Also, correct me if I'm wrong, but one implements these Key functions in key package, then every function that somehow utilizes custom object would need type switch to determine whether it's `AWSClusterConfig`, `AzureClusterConfig` or `KVMClusterConfig`. I personally would find this ugly especially when there's language feature (interface + inversion) to implement this cleanly.

Do I get the versioning wrong here and hence get driven to wrong direction?

Reply:
Versioning has nothing to do with object versions. It's about dispatching objects with versionBundle.Version set. See https://github.com/giantswarm/kvm-operator/blob/925326e428f54ebe54bff5d66a1eb68f87984c4f/service/kvmconfig/framework.go#L101-L201

But now I see your problem. I think it would be less intrusive to add an extractor to as a resource dependency. Like:

```go
type Config struct {
	ToClusterConfigFunc func(obj interface{}) (*ClusterConfig, error)
}
```

That can be provided by `{aws,azure,kvm}clusterconfig`.

Reply:
> a new version is created and implementation that uses it upgrades when suitable

Yeah sure. I understand this. It is just that we now create just another exception that works in just another way and I already see the situation in which an SRE or an solutions engineer or somebody from the community tries to see light through all of this and has to implement something and out of a sudden there is just another concept that makes it just a little bit harder to understand what and how to do it and then you have to explain this and we have a lot of discussions again. I think it is important to consider all kinds of people who have to deal with this here. I comprehend what you want to do here but IMO it is not making it better for the company nor the community and will just lead to more pain in certain places (even though you are kind of right <3). We could just keep everything as it is and be fine but I made my point now several times. So do what you think is right. 

> Versioning has nothing to do with object versions.

IMO it has. Depending on the case you have to deal with updating certain things and then the whole versioning is just getting more complicated.

Reply:
Ok. Tim has a good point on various people having need to understand the mechanisms here. I'll go with Pawel's suggestion and use extraction function together with key package.

Reply:
Just to be on the same page my concerns are more going towards the place where we put the resources and the implications that will have for versioning and people who have to do the leg work. It is also that it will become even more complicated for any kind of tooling that is supposed to help us here to actually help us because we have to make them even smarter.

Reply:
Versioning is never easy. I think we can see it from Go's current situation as well. I think that currently there's no perfect solution here. For what I see (please correct me if I miss something) we have following options:
 - Use the current approach and store shared resources under `/pkg`. As a general structure this is well understood within Go community.
 - Store shared resources under `/services/shared/v1`. This could follow current path structure for scripts but is not necessarily as clear to denote shared nature of resources.
 - Duplicate resource implementation for each provider. I personally consider this as a very last resort. Code maintenance would definitely become a nightmare due to requirement of making same modification in multiple places (while eventually still drifting apart due to versioning differences). This would also be quite an appalling find for anyone who doesn't know the origins of design decision.

If we have difficulties with living with this approach, I think this can be relatively easily changed to something else and at that point we need to revisit our versioning practices in general if there's anything we can do to improve here.